File: Lib/Telerik/Examples/QuickStart/Scheduler/CustomAppointmentDialog/CustomEditAppointmentDialog.cs
using System;
using System.Windows.Forms;
using Telerik.WinControls.UI.Scheduler.Dialogs;
using Telerik.WinControls.UI;
using Telerik.WinControls.Enumerations;
using Telerik.WinControls;

namespace Telerik.Examples.WinControls.Scheduler.CustomAppointmentDialog
{
    public partial class CustomEditAppointmentDialog : RadSchedulerDialog, IEditAppointmentDialog
    {
        #region fields

        private static readonly TimeSpan EndOfDayTimeSpan = new TimeSpan(23, 59, 59);
        private static readonly TimeSpan StartOfDayTimeSpan = new TimeSpan(0, 0, 0);

        private IEvent appointment = null;
        private ISchedulerData schedulerData = null;
        protected IOpenRecurringAppointmentDialog openRecurringAppointmentDialog = null;

        private IEvent recurringAppointment = null;
        private bool saveRecurringAppointment = false;

        #endregion

        #region constructors

        public CustomEditAppointmentDialog()
        {
            InitializeComponent();

            WireEvents();
        }

        #endregion

        #region IEditAppointmentDialog Members

        public bool EditAppointment(IEvent appointment, ISchedulerData schedulerData)
        {
            this.appointment = appointment;
            this.schedulerData = schedulerData;

            // show OpenRecurringAppointmentDialog
            if (this.appointment != null && this.appointment.MasterEvent != null)
            {
                if (this.openRecurringAppointmentDialog == null)
                {
                    this.openRecurringAppointmentDialog = new OpenRecurringAppointmentDialog();
                }
                this.openRecurringAppointmentDialog.ThemeName = this.ThemeName;
                this.openRecurringAppointmentDialog.EventName = appointment.Summary;
                DialogResult result = this.openRecurringAppointmentDialog.ShowDialog();
                if (result != DialogResult.OK)
                {
                    return false;
                }

                bool editOccurrence = this.openRecurringAppointmentDialog.EditOccurrence;
                if (!editOccurrence)
                {
                    this.appointment = this.appointment.MasterEvent as OutlookLikeAppointment;
                }
            }

            return true;
        }

        public void ShowRecurrenceDialog()
        {
            // initialize the recurringAppointment as OutlookLikeAppointment
            if (this.recurringAppointment == null)
            {
                this.recurringAppointment = new OutlookLikeAppointment();
            }

            // set settings
            this.recurringAppointment.Start = this.GetAppointmentStart();
            this.recurringAppointment.End = this.GetAppointmentEnd();
            if (!this.RecurrenceSettingsShouldBeSaved() && this.appointment.RecurrenceRule != null)
            {
                this.recurringAppointment.RecurrenceRule = this.appointment.RecurrenceRule.Clone();
            }

            // show EditRecurrenceDialog
            IEditRecurrenceDialog editRecurrenceDialog = new EditRecurrenceDialog(recurringAppointment, this.schedulerData);
            editRecurrenceDialog.ThemeName = this.ThemeName;
            DialogResult result = editRecurrenceDialog.ShowDialog();
            if (result == DialogResult.OK)
            {
                this.saveRecurringAppointment = true;

                this.SetStartDateAndTime(this.recurringAppointment.Start);
                this.SetEndDateAndTime(this.recurringAppointment.End);

                ApplyRecurrenceBehavior(this.recurringAppointment);
            }
        }

        #endregion

        #region events

        protected override void OnLoad(EventArgs e)
        {
            base.OnLoad(e);

            this.saveRecurringAppointment = false;
            this.recurringAppointment = null;

            // load appointment settings and controls data
            this.LoadStatuses();
            this.LoadSettingsFromEvent(this.appointment);
            this.ApplyRecurrenceBehavior(this.appointment);

            this.btnRecurrence.Enabled = (this.appointment.MasterEvent == null);
        }

        protected override void OnShown(EventArgs e)
        {
            base.OnShown(e);

            this.txtSubject.Focus();
        }

        private void btnSave_Click(object sender, EventArgs e)
        {
            this.ApplySettingsToEvent(this.appointment);

            this.DialogResult = DialogResult.OK;
            this.Close();
        }

        private void btnDelete_Click(object sender, EventArgs e)
        {
            if (this.schedulerData != null)
            {
                this.schedulerData.GetEventStorage().Remove(this.appointment);
            }

            this.DialogResult = DialogResult.Abort;
            this.Close();
        }

        private void btnRecurrence_Click(object sender, EventArgs e)
        {
            this.ShowRecurrenceDialog();
        }

        #endregion

        #region private methods

        private void LoadStatuses()
        {
            if (this.schedulerData == null)
            {
                return;
            }

            ISchedulerStorage<IAppointmentStatusInfo> statusStorage = this.schedulerData.GetStatusStorage();
            this.cmbShowTimeAs.BeginUpdate();
            this.cmbShowTimeAs.Items.Clear();
            foreach (IAppointmentStatusInfo status in statusStorage)
            {
                this.cmbShowTimeAs.Items.Add(new RadListDataItem(status.DisplayName, status.Id));
            }

            if (this.cmbShowTimeAs.Items.Count > 0)
            {
                this.cmbShowTimeAs.SelectedIndex = 0;
            }

            this.cmbShowTimeAs.EndUpdate();
        }


        private void ApplySettingsToEvent(IEvent targetEvent)
        {
            // appointment settings
            targetEvent.Summary = this.txtSubject.Text;
            targetEvent.Location = this.txtLocation.Text;
            targetEvent.Description = this.txtDescription.Text;

            object selectedStatus = this.cmbShowTimeAs.SelectedValue;
            targetEvent.StatusId = (selectedStatus != null) ? (int)selectedStatus : 1;

            targetEvent.Start = this.GetAppointmentStart();
            targetEvent.End = this.GetAppointmentEnd();

            if (this.RecurrenceSettingsShouldBeSaved())
            {
                targetEvent.RecurrenceRule = this.recurringAppointment.RecurrenceRule;
            }

            // if we are editing an occurrence, add an exception
            if (targetEvent.MasterEvent != null) 
            {
                if (!targetEvent.MasterEvent.Exceptions.Contains(targetEvent))
                {
                    if (targetEvent.RecurrenceId == null)
                    {
                        targetEvent.RecurrenceId = targetEvent.Start;
                    }
                    targetEvent.MasterEvent.Exceptions.Add(targetEvent);
                }
            }

            // OutlookLike settings
            OutlookLikeAppointment outlookAppointment = targetEvent as OutlookLikeAppointment;
            if (outlookAppointment == null)
                return;
            outlookAppointment.Email = this.txtEmail.Text;
        }

        private void LoadSettingsFromEvent(IEvent sourceEvent)
        {
            // appointment settings
            this.txtSubject.Text = sourceEvent.Summary;
            this.txtLocation.Text = sourceEvent.Location;
            this.txtDescription.Text = sourceEvent.Description;

            this.cmbShowTimeAs.SelectedValue = sourceEvent.StatusId;

            this.SetStartDateAndTime(sourceEvent.Start);
            this.SetEndDateAndTime(sourceEvent.End);

            double totalHours = sourceEvent.Duration.TotalHours;
            bool isAllDay = totalHours >= 23.99 && Math.Round(totalHours) % 24 == 0;
            this.chkAllDay.ToggleState = isAllDay ? ToggleState.On : ToggleState.Off;

            if (this.appointment != null && this.schedulerData != null)
            {
                this.btnDelete.Enabled = schedulerData.GetEventStorage().Contains(this.appointment) && sourceEvent.AllowDelete;
            }

            this.SetTimePickersEnabledState(isAllDay);

            // OutlookLike settings
            OutlookLikeAppointment outlookAppointment = sourceEvent as OutlookLikeAppointment;
            if (outlookAppointment == null)
            {
                this.txtEmail.Text = String.Empty;
                return;
            }
            this.txtEmail.Text = outlookAppointment.Email;
        }


        private void ApplyRecurrenceBehavior(IEvent appointment)
        {
            bool isOutlookBehavior = false;
            Telerik.WinControls.UI.RadScheduler scheduler = this.schedulerData as Telerik.WinControls.UI.RadScheduler;

            if (scheduler != null)
            {
                if (appointment != null && appointment.RecurrenceRule != null)
                {
                    isOutlookBehavior = true;
                }
            }

            this.dateStart.Enabled = !isOutlookBehavior;
            this.dateEnd.Enabled = !isOutlookBehavior;
            this.timeStart.Enabled = !isOutlookBehavior && !this.chkAllDay.Checked;
            this.timeEnd.Enabled = !isOutlookBehavior && !this.chkAllDay.Checked;
            this.chkAllDay.Visible = !isOutlookBehavior;
        }

        private bool RecurrenceSettingsShouldBeSaved()
        {
            return this.saveRecurringAppointment && this.recurringAppointment != null;
        }

        private void SetTimePickersEnabledState(bool allDay)
        {
            bool enabled = !allDay;
            this.timeStart.Enabled = enabled;
            this.timeEnd.Enabled = enabled;
        }


        private DateTime GetAppointmentStart()
        {
            DateTime startDate = this.dateStart.Value.Date;

            if (!(this.chkAllDay.ToggleState == ToggleState.On))
            {
                TimeSpan startTime = this.timeStart.Value.TimeOfDay;
                startDate = startDate.Add(startTime);
            }

            return startDate;
        }

        private DateTime GetAppointmentEnd()
        {
            DateTime endDate = this.dateEnd.Value.Date;
            TimeSpan endTime = this.timeEnd.Value.TimeOfDay;

            bool isAllDay = this.chkAllDay.ToggleState == ToggleState.On;
            if (!isAllDay)
            {
                endDate = endDate.Add(endTime);
            }
            else
            {
                endDate = endDate.Add(CustomEditAppointmentDialog.EndOfDayTimeSpan);
            }

            return endDate;
        }

        private void SetStartDateAndTime(DateTime start)
        {
            this.dateStart.Value = start;
            this.timeStart.Value = start;
        }

        private void SetEndDateAndTime(DateTime end)
        {
            this.dateEnd.Value = end;
            this.timeEnd.Value = end;
        }

        #endregion

        protected override void OnThemeChanged()
        {
            base.OnThemeChanged();

            ThemeResolutionService.ApplyThemeToControlTree(this, this.ThemeName);
        }

        private void WireEvents()
        {
            this.btnSave.Click += new System.EventHandler(this.btnSave_Click);
            this.btnDelete.Click += new System.EventHandler(this.btnDelete_Click);
            this.btnRecurrence.Click += new System.EventHandler(this.btnRecurrence_Click);
        }
    }
}

File: Lib/Telerik/Examples/QuickStart/Scheduler/CustomAppointmentDialog/CustomEditAppointmentDialog.vb
Imports Telerik.WinControls.UI.Scheduler.Dialogs
Imports Telerik.WinControls.UI
Imports Telerik.WinControls.Enumerations
Imports Telerik.WinControls

Namespace Telerik.Examples.WinControls.Scheduler.CustomAppointmentDialog
	Partial Public Class CustomEditAppointmentDialog
		Inherits RadSchedulerDialog
		Implements IEditAppointmentDialog
		#Region "fields"

		Private Shared ReadOnly EndOfDayTimeSpan As New TimeSpan(23, 59, 59)
		Private Shared ReadOnly StartOfDayTimeSpan As New TimeSpan(0, 0, 0)

		Private appointment As IEvent = Nothing
		Private schedulerData As ISchedulerData = Nothing
		Protected openRecurringAppointmentDialog As IOpenRecurringAppointmentDialog = Nothing

		Private recurringAppointment As IEvent = Nothing
		Private saveRecurringAppointment As Boolean = False

		#End Region

		#Region "constructors"

		Public Sub New()
			InitializeComponent()

			WireEvents()
		End Sub

		#End Region

		#Region "IEditAppointmentDialog Members"

        Public Function EditAppointment(ByVal appointment As IEvent, ByVal schedulerData As ISchedulerData) As Boolean Implements IEditAppointmentDialog.EditAppointment
            Me.appointment = appointment
            Me.schedulerData = schedulerData

            ' show OpenRecurringAppointmentDialog
            If Me.appointment IsNot Nothing AndAlso Me.appointment.MasterEvent IsNot Nothing Then
                If Me.openRecurringAppointmentDialog Is Nothing Then
                    Me.openRecurringAppointmentDialog = New OpenRecurringAppointmentDialog()
                End If
                Me.openRecurringAppointmentDialog.ThemeName = Me.ThemeName
                Me.openRecurringAppointmentDialog.EventName = appointment.Summary
                Dim result As DialogResult = Me.openRecurringAppointmentDialog.ShowDialog()
                If result <> DialogResult.OK Then
                    Return False
                End If

                Dim editOccurrence As Boolean = Me.openRecurringAppointmentDialog.EditOccurrence
                If Not editOccurrence Then
                    Me.appointment = TryCast(Me.appointment.MasterEvent, OutlookLikeAppointment)
                End If
            End If

            Return True
        End Function

        Public Sub ShowRecurrenceDialog() Implements IEditAppointmentDialog.ShowRecurrenceDialog
            ' initialize the recurringAppointment as OutlookLikeAppointment
            If Me.recurringAppointment Is Nothing Then
                Me.recurringAppointment = New OutlookLikeAppointment()
            End If

            ' set settings
            Me.recurringAppointment.Start = Me.GetAppointmentStart()
            Me.recurringAppointment.End = Me.GetAppointmentEnd()
            If (Not Me.RecurrenceSettingsShouldBeSaved()) AndAlso Me.appointment.RecurrenceRule IsNot Nothing Then
                Me.recurringAppointment.RecurrenceRule = Me.appointment.RecurrenceRule.Clone()
            End If

            ' show EditRecurrenceDialog
            Dim editRecurrenceDialog As IEditRecurrenceDialog = New EditRecurrenceDialog(recurringAppointment, Me.schedulerData)
            editRecurrenceDialog.ThemeName = Me.ThemeName
            Dim result As DialogResult = editRecurrenceDialog.ShowDialog()
            If result = DialogResult.OK Then
                Me.saveRecurringAppointment = True

                Me.SetStartDateAndTime(Me.recurringAppointment.Start)
                Me.SetEndDateAndTime(Me.recurringAppointment.End)

                ApplyRecurrenceBehavior(Me.recurringAppointment)
            End If
        End Sub

        Public Overloads Function ShowDialog() As System.Windows.Forms.DialogResult Implements IEditAppointmentDialog.ShowDialog
            Return MyBase.ShowDialog()
        End Function

        Private Shadows Event Shown As System.EventHandler Implements IEditAppointmentDialog.Shown

#End Region

#Region "events"

        Protected Overloads Overrides Sub OnLoad(ByVal e As EventArgs)
            MyBase.OnLoad(e)

            Me.saveRecurringAppointment = False
            Me.recurringAppointment = Nothing

            ' load appointment settings and controls data
            Me.LoadStatuses()
            Me.LoadSettingsFromEvent(Me.appointment)
            Me.ApplyRecurrenceBehavior(Me.appointment)

            Me.btnRecurrence.Enabled = (Me.appointment.MasterEvent Is Nothing)
        End Sub

        Protected Overrides Sub OnShown(e As EventArgs)
            MyBase.OnShown(e)

            Me.txtSubject.Focus()
        End Sub

        Private Sub btnSave_Click(ByVal sender As Object, ByVal e As EventArgs)
            Me.ApplySettingsToEvent(Me.appointment)

            Me.DialogResult = DialogResult.OK
            Me.Close()
        End Sub

        Private Sub btnDelete_Click(ByVal sender As Object, ByVal e As EventArgs)
			If Me.schedulerData IsNot Nothing Then
				Me.schedulerData.GetEventStorage().Remove(Me.appointment)
			End If

			Me.DialogResult = DialogResult.Abort
			Me.Close()
		End Sub

		Private Sub btnRecurrence_Click(ByVal sender As Object, ByVal e As EventArgs)
			Me.ShowRecurrenceDialog()
		End Sub

		#End Region

		#Region "private methods"

		Private Sub LoadStatuses()
			If Me.schedulerData Is Nothing Then
				Return
			End If

			Dim statusStorage As ISchedulerStorage(Of IAppointmentStatusInfo) = Me.schedulerData.GetStatusStorage()
			Me.cmbShowTimeAs.BeginUpdate()
			Me.cmbShowTimeAs.Items.Clear()
			For Each status As IAppointmentStatusInfo In statusStorage
				Me.cmbShowTimeAs.Items.Add(New RadListDataItem(status.DisplayName, status.Id))
			Next status

			If Me.cmbShowTimeAs.Items.Count > 0 Then
				Me.cmbShowTimeAs.SelectedIndex = 0
			End If

			Me.cmbShowTimeAs.EndUpdate()
		End Sub


		Private Sub ApplySettingsToEvent(ByVal targetEvent As IEvent)
			' appointment settings
			targetEvent.Summary = Me.txtSubject.Text
			targetEvent.Location = Me.txtLocation.Text
			targetEvent.Description = Me.txtDescription.Text

			Dim selectedStatus As Object = Me.cmbShowTimeAs.SelectedValue
			targetEvent.StatusId = If((selectedStatus IsNot Nothing), CInt(Fix(selectedStatus)), 1)

			targetEvent.Start = Me.GetAppointmentStart()
			targetEvent.End = Me.GetAppointmentEnd()

			If Me.RecurrenceSettingsShouldBeSaved() Then
				targetEvent.RecurrenceRule = Me.recurringAppointment.RecurrenceRule
			End If

			' if we are editing an occurrence, add an exception
			If targetEvent.MasterEvent IsNot Nothing Then
				If Not targetEvent.MasterEvent.Exceptions.Contains(targetEvent) Then
					If targetEvent.RecurrenceId Is Nothing Then
						targetEvent.RecurrenceId = targetEvent.Start
					End If
					targetEvent.MasterEvent.Exceptions.Add(targetEvent)
				End If
			End If

			' OutlookLike settings
			Dim outlookAppointment As OutlookLikeAppointment = TryCast(targetEvent, OutlookLikeAppointment)
			If outlookAppointment Is Nothing Then
				Return
			End If
			outlookAppointment.Email = Me.txtEmail.Text
		End Sub

		Private Sub LoadSettingsFromEvent(ByVal sourceEvent As IEvent)
			' appointment settings
			Me.txtSubject.Text = sourceEvent.Summary
			Me.txtLocation.Text = sourceEvent.Location
			Me.txtDescription.Text = sourceEvent.Description

			Me.cmbShowTimeAs.SelectedValue = sourceEvent.StatusId

			Me.SetStartDateAndTime(sourceEvent.Start)
			Me.SetEndDateAndTime(sourceEvent.End)

			Dim totalHours As Double = sourceEvent.Duration.TotalHours
			Dim isAllDay As Boolean = totalHours >= 23.99 AndAlso Math.Round(totalHours) Mod 24 = 0
			Me.chkAllDay.ToggleState = If(isAllDay, ToggleState.On, ToggleState.Off)

			If Me.appointment IsNot Nothing AndAlso Me.schedulerData IsNot Nothing Then
				Me.btnDelete.Enabled = schedulerData.GetEventStorage().Contains(Me.appointment) AndAlso sourceEvent.AllowDelete
			End If

			Me.SetTimePickersEnabledState(isAllDay)

			' OutlookLike settings
			Dim outlookAppointment As OutlookLikeAppointment = TryCast(sourceEvent, OutlookLikeAppointment)
			If outlookAppointment Is Nothing Then
				Me.txtEmail.Text = String.Empty
				Return
			End If
			Me.txtEmail.Text = outlookAppointment.Email
		End Sub


		Private Sub ApplyRecurrenceBehavior(ByVal appointment As IEvent)
			Dim isOutlookBehavior As Boolean = False
			Dim scheduler As Telerik.WinControls.UI.RadScheduler = TryCast(Me.schedulerData, Telerik.WinControls.UI.RadScheduler)

			If scheduler IsNot Nothing Then
				If appointment IsNot Nothing AndAlso appointment.RecurrenceRule IsNot Nothing Then
					isOutlookBehavior = True
				End If
			End If

			Me.dateStart.Enabled = Not isOutlookBehavior
			Me.dateEnd.Enabled = Not isOutlookBehavior
			Me.timeStart.Enabled = (Not isOutlookBehavior) AndAlso Not Me.chkAllDay.Checked
			Me.timeEnd.Enabled = (Not isOutlookBehavior) AndAlso Not Me.chkAllDay.Checked
			Me.chkAllDay.Visible = Not isOutlookBehavior
		End Sub

		Private Function RecurrenceSettingsShouldBeSaved() As Boolean
			Return Me.saveRecurringAppointment AndAlso Me.recurringAppointment IsNot Nothing
		End Function

		Private Sub SetTimePickersEnabledState(ByVal allDay As Boolean)
			Dim enabled As Boolean = Not allDay
			Me.timeStart.Enabled = enabled
			Me.timeEnd.Enabled = enabled
		End Sub


		Private Function GetAppointmentStart() As Date
			Dim startDate As Date = Me.dateStart.Value.Date

			If Not(Me.chkAllDay.ToggleState = ToggleState.On) Then
				Dim startTime As TimeSpan = Me.timeStart.Value.TimeOfDay
				startDate = startDate.Add(startTime)
			End If

			Return startDate
		End Function

		Private Function GetAppointmentEnd() As Date
			Dim endDate As Date = Me.dateEnd.Value.Date
			Dim endTime As TimeSpan = Me.timeEnd.Value.TimeOfDay

			Dim isAllDay As Boolean = Me.chkAllDay.ToggleState = ToggleState.On
			If Not isAllDay Then
				endDate = endDate.Add(endTime)
			Else
				endDate = endDate.Add(CustomEditAppointmentDialog.EndOfDayTimeSpan)
			End If

			Return endDate
		End Function

		Private Sub SetStartDateAndTime(ByVal start As Date)
			Me.dateStart.Value = start
			Me.timeStart.Value = start
		End Sub

		Private Sub SetEndDateAndTime(ByVal [end] As Date)
			Me.dateEnd.Value = [end]
			Me.timeEnd.Value = [end]
		End Sub

		#End Region

		Protected Overrides Sub OnThemeChanged()
			MyBase.OnThemeChanged()

			ThemeResolutionService.ApplyThemeToControlTree(Me, Me.ThemeName)
		End Sub

		Private Sub WireEvents()
			AddHandler btnSave.Click, AddressOf btnSave_Click
			AddHandler btnDelete.Click, AddressOf btnDelete_Click
			AddHandler btnRecurrence.Click, AddressOf btnRecurrence_Click
		End Sub
	End Class
End Namespace